cloudinit/socket.py

```diff
@@ -0,0 +1,164 @@
+                pass
+            with sync("stage 2"):
+                pass
+        """
+        if stage not in self.sockets:
+            raise ValueError(f"Invalid stage name: {stage}")
+        self.stage = stage
+        return self
+
+    def __enter__(self):
+        """Wait until a message has been received on this stage's socket.
+
+        Once the message has been received, enter the context.
+        """
+        if os.isatty(sys.stdin.fileno()):
+            LOG.info("Not syncing, stdin is a tty.")
+            return
+        sd_notify(
+            "STATUS=Waiting on external services to "
+            f"complete ({self.stage} stage)"
+        )
+        start_time = time.monotonic()
+        # block until init system sends us data
+        # the first value returned contains a message from the init system
+        #     (should be "start")
+        # the second value contains the path to a unix socket on which to
+        #     reply, which is expected to be /path/to/{self.stage}-return.sock
+        sock = self.sockets[self.stage]
+        chunk, self.remote = sock.recvfrom(5)
+
+        if b"start" != chunk:
+            # The protocol expects to receive a command "start"
+            self.__exit__(None, None, None)
+            raise ValueError(f"Received invalid message: [{str(chunk)}]")
+        elif f"{DEFAULT_RUN_DIR}/share/{self.stage}-return.sock" != str(
+            self.remote
+        ):
+            # assert that the return path is in a directory with appropriate
+            # permissions
+            self.__exit__(None, None, None)
+            raise ValueError(f"Unexpected path to unix socket: {self.remote}")
+
+        total = time.monotonic() - start_time
+        time_msg = f"took {total: .3f}s " if total > 0.1 else ""
+        sd_notify(f"STATUS=Running ({self.stage} stage)")
+        LOG.debug("sync(%s): synchronization %scomplete", self.stage, time_msg)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Notify the socket that this stage is complete."""
+        message = "done"
+        systemd_exit_code = "0"
+        if exc_type:
+            # handle exception thrown in context
+            systemd_exit_code = "1"
+            status = f"{repr(exc_val)} in {exc_tb.tb_frame}"
+            message = (
+                'fatal error, run "systemctl cloud-init.service" for more '
+                "details"
+            )
+            if not self.first_exception:
+                self.first_exception = message
+            LOG.fatal(status)
+            sd_notify(f"STATUS={status}")
+
+        sock = self.sockets[self.stage]
+        sock.connect(self.remote)
+
+        # the returned message will be executed in a subshell
+        # hardcode this message rather than sending a more informative message
+        # to avoid having to sanitize inputs (to prevent escaping the shell)
+        sock.sendall(f"echo '{message}'; exit {systemd_exit_code};".encode())
```

Comment:
I had an entire comment here about how I don't like this, but once I made it to the `.service` files, I see what's going on, and it kind of makes sense. It's clever. Maybe too clever? Will we miss having process names in `ps` output that reflect what the process is doing rather than the program that's doing it?

Reply:
> I had an entire comment here about how I don't like this, but once I made it to the .service files, I see what's going on, and it kind of makes sense. It's clever. Maybe too clever?

I don't love it myself, but I prefer this solution to the alternatives. I think that introducing a compiled language for performance-critical sections like this might be the best solution long term, but that's also probably bigger than this PR. I've dealt with enough shell parsing bugs to want to avoid the possibility of introducing more in this PR. The alternative to shell and compiled language is a Python interpreter which.... is exactly what we're trying to avoid given the performance requirements. Or rather than doing anything at all we could just not pass along error messages and exit codes, however that would be much less user-friendly for anyone debugging cloud-init.

If / when cloud-init introduces a compiled language I'd want this to be one of the top things to replace, but I think that doing that is a much bigger effort.

> Will we miss having process names in ps output that reflect what the process is doing rather than the program that's doing it?

I'm not too worried about that, since checking the PGID (the logical next step) should point anyone debugging in the right direction.

Reply:
> If / when cloud-init introduces a compiled language I'd want this to be one of the top things to replace, but I think that doing that is a much bigger effort.

Do you think it is worth creating a GH issue tracking such a wishlist for things to be replaced by a compiled language?

Reply:
> > If / when cloud-init introduces a compiled language I'd want this to be one of the top things to replace, but I think that doing that is a much bigger effort.
> 
> Do you think it is worth creating a GH issue tracking such a wishlist for things to be replaced by a compiled language?

[done](https://github.com/canonical/cloud-init/issues/5552)